### DISCUSSION TITLE
Trying to add a feature to the "MusicWheelItem grades" that displays a symbol if the player got a full combo...

### DISCUSSION THREAD

**Discussion by deathzero021** [stepmania/stepmania#2111] [Q&A] [answered]
I'm not sure where the best place to ask such questions would be, most Stepmania related forums are pretty dead these days and I suppose this is a bit of a technical question. Here's the script:

```
local grades = {
	Grade_Tier01 = 0;
	Grade_Tier02 = 1;
	Grade_Tier03 = 2;
	Grade_Tier04 = 3;
	Grade_Tier05 = 4;
	Grade_Tier06 = 5;
	Grade_Tier07 = 6;
	Grade_Failed = 7;
	Grade_None = 8;
};

local t = LoadFont("Common Normal") .. {
	InitCommand=cmd(zoom,0.75;shadowlength,1;strokecolor,Color("Black"));
	ShowCommand=cmd(stoptweening;bounceend,0.15;zoomy,0.75);
	HideCommand=cmd(stoptweening;bouncebegin,0.15;zoomy,0);
	GameCommand=function(self,params)
		local steps = GetSteps();
	end;
	
	SetGradeCommand=function(self,params)
		local pnPlayer = params.PlayerNumber;
		local sGrade = params.Grade or 'Grade_None';
		local gradeString = THEME:GetString("Grade",string.sub(sGrade,7));
		
		local star = "";
		local song = params.Song;
		
		local playerProfile = GetPlayerOrMachineProfile(pnPlayer); 
		local playerHighScoreList = playerProfile:GetHighScoreListIfExists(song, steps);
		if playerHighScoreList ~= nil then
			local playerHighScore = playerHighScoreList:GetHighestScoreOfName("EVNT");
			local playerStageAward = playerHighScore:GetStageAward();
			if playerStageAward ~= nil then
				star = "★";
			end;
		end;
		
		self:settext(string.format("%s%s", gradeString, star));
		self:diffuse(PlayerColor(pnPlayer));
		self:diffusetopedge(BoostColor(PlayerColor(pnPlayer),1.5));
		self:strokecolor(BoostColor(PlayerColor(pnPlayer),0.25));
		
	end;
};

return t;
```

I got the error:
`calling GetHighScoreListIfExists on bad self (Song expected, got nil)`

I can't figure out what's wrong. I'm not experienced in lua at all, but I did read some guides and have some understanding of the syntax. I've done a few other minor modifications to the default theme as well without issue. I followed a pdf that showed all the lua bindings for Stepmania 5, though I'm not sure which exact version that guide was written for. So I think I'm calling the right methods from the appropriate object but evidently not. Any help would be greatly appreciated! As well as any suggestions for other places I could post these types of questions.

**Comment by Inorizushi** [comment, marked as answer]
You'd be better off adding this code to MusicWheelItem Song NormalPart and stubbing out MWI grades.

I have no idea but you likely can't get the song in MWI grades, so you can easily replace them with this code block.

```lua
local t = Def.ActorFrame{};

--Use a for loop so you don't end up copy+pasting this code twice for each player.
for _, pn in pairs(GAMESTATE:GetEnabledPlayers()) do
	t[#t+1] = Def.BitmapText{
		Font="Common Normal",
		InitCommand=function(s) s:zoom(0.75):shadowlength(1)
			--flip flop to save typing/lines. This changes the position depending on which player the actor is for.
			:xy(pn==PLAYER_1 and -100 or 100,pn==PLAYER_1 and -100 or 100)
		end,
		SetCommand=function(s,p)
			--We make our variables here since we'll be modifying them later on.
			local star = ""
			local gString = ""
			local pG = ""
			--bool flip flop to save lines again
			local song = GAMESTATE:IsCourseMode() and p.Course or p.Song
			local steps = GAMESTATE:IsCourseMode() and GAMESTATE:GetCurrentTrail(pn) or GAMESTATE:GetCurrentSteps(pn)
			if song and steps then
				local profile = GetPlayerOrMachineProfile(pn)
				local pHSL = profile:GetHighScoreListIfExists(song,steps)
				if pHSL then
					local pS = pHSL:GetHighScores()
					if pS[1] then
						local pSA = pS[1]:GetStageAward();
						pG = pS[1]:GetGrade()
						gString = THEME:GetString("Grade",string.sub(pG,7))
						if pSA then
							star = "★"
						end
					end
					s:settext(string.format("%s%s", gString, star))
					:diffuse(PlayerColor(pn)):diffusetopedge(BoostColor(PlayerColor(pn),1.5))
					:strokecolor(BoostColor(PlayerColor(pn),0.25))
				end
			end
		end,
		CurrentSongChangedMessageCommand=cmd(queuecommand,"Set");
		--we use sesub here so we have a single line but it still queues the command for either player.
		[sesub("CurrentStepsChanged%MessageCommand",pn)]=cmd(queuecommand,"Set");
	}
end

return t;
```

**Reply by deathzero021** [reply]
Hey, thanks for the reply!

So it seems to be working mostly, but I get this error:
`attempt to call global sesub (a nil value)`

Commenting out the sesub line avoids the error and gets results. But there is a problem, most likely related to that line not working. It doesn't update the grades when the difficulty changes, unless you move up/down in the menu. Also no grades are shown when the menu first opens up. 

Also I get error messages when both players are enabled, and no grade for P2, though that could be related to a different script I wrote... seeing as the score % also doesn't show up for P2, however those problems don't exist when playing as P2 alone, so it only shows up when both players are present. Weird. 

Edit: I replaced the sesub line with this:
`CurrentStepsP1ChangedMessageCommand=cmd(queuecommand,"Set");`
No errors now. However the issue of it not updating the grade when the steps change still persists.

**Reply by Inorizushi** [reply]
I've looked over the code again and suddenly sesub isn't working on my copy (It's likely a function that was in the theme I was using before switching to the one I'm testing this with)

Be sure to add `CurrentStepsChangedP2MessageCommand=cmd(queuecommand,"Set");` as well so the P2 grade gets updated too.

**Reply by deathzero021** [reply]
Understood. Any clues as to why the grade isn't updating when changing Steps? It also doesn't update when opening/closing a group, so it'll display grades for the previous group even when you open a new one, or close the previous group. 

Only thing I can think of is that maybe "MusicWheelItem Song NormalPart" isn't intended to update when steps change. Since it's original intended purpose is just to display an image, perhaps that's why queuecommand won't work? I've tried playcommand too but that gives an error, something about p being nil.

**Reply by Inorizushi** [reply]
No that's not the issue, there's plenty of themes out there that update the actor when steps change and it's working fine over here. What version of sm5 are you using and by any chance can I get the theme so I can look a bit further?

**Reply by deathzero021** [reply]
Stepmania 5.0.12 is the version I'm running. The theme is just the default for that version, with colors modified and StepsDisplay modified to display a different difficulty value (calculated based on the radar values.) Though I can zip it if necessary.

**Reply by Inorizushi** [reply]
Ah ok, 5.0.12 is a bit weird when it comes to the StepsChangedCommand, lemme update the code so it works properly

**Reply by Inorizushi** [reply]
alright i've got some code written out but it currently isn't updating until you change songs (probably because of 5.0.12 issues)

**Reply by deathzero021** [reply]
Haha figures. Should I abandon 5.0.12 than? Would the 5.0.12 theme run on 5.1? I suppose I'll test that out for now.

**Reply by Horsey-** [reply]
I’ve been using 5.1 and 5.3, with zero issues so far, for years now. 5.0.12 just exists on the website because there was uncertainty with the 5.1/5.2 split (and other drama). 5.3 is being developed in a closed manner but has plans to be properly merged with 5.1 sometime in the future.  5.3 has the added benefit of being developed with modern coding practices, so if you’re running on modern hardware, 5.3 will give you a huge performance boost.

**Reply by deathzero021** [reply]
I just installed 5.1 and it seems the 5.0.12 theme does work without any immediate issues. Also it fixed this issue, the grades now update appropriately when steps or song changes. So I can consider this thread solved now. Big thank you to Inorizushi!

Only downside I've noticed though is that my scores didn't transfer over :\
I copied the profile, and it loaded the profile, including my options and such, but no scores. I guess the xml formatting changed or something between versions? Kind of a bummer haha. 

Also, on an unrelated note: when a 2nd player joins while in the music wheel menu, the 2nd player won't display grades/scores (even the score% doesn't draw anything, though the frame and steps value does.) If I play a song than go back to the music wheel, the 2nd player's scores and grades load and display properly. Similarly, if I enter the music wheel menu with both players already joined from the main menu, than it'll work correctly too. So it's only when the 2nd player joins in after the 1st player has already entered the music wheel. Could be an issue on my side of things though but not sure what's causing that. It's not really an important issue though haha.

**Reply by Horsey-** [reply]
5.0.12, 5.1, 5.3 scores are all interchangeable with each other (I have upgraded to 5.3 after using 5.1). You just need to be careful when transferring the scores to keep the machine profile and local profiles straight.

**Reply by deathzero021** [reply]
Ahh, i see what the problem is. It's because of the "additionalsongs" in the directory name of the songs. Weird that it won't load a score of the same song if the song path is using the additionalsongs preference. I guess I'll just move the songs folder over than. Thanks for the help Horsey-.

Edit: Upon further testing, it seems it still doesn't update perfectly. In most cases it seems to work as intended. But for songs that don't have a certain difficulty, it doesn't seem to update when switching to a difficulty it doesn't have steps for. In this case, it'll just continue displaying the last letter grade that it was showing before the difficulty changed. This isn't always the case though, for example:

If my difficulty is set on Hard, my current song only has Expert, the current song will display the Expert grade, but if I move to the next song, and the next song does have Hard, than the previous song won't update it's grade (it should show nothing since that song doesn't have a Hard steps) it'll still show it's Expert grade despite the current song and difficulty is Hard.

I'm not sure if there's gonna be any fix for this though. Might as well consider this the best it can get haha. I guess the `CurrentStepsP1ChangedMessageCommand=cmd(queuecommand,"Set"); `only gets called when the player manually changes difficulty, and not when the game forces the difficulty change due to a song not having the set difficulty.

**Reply by Inorizushi** [reply]
as for the music wheel not updating when another player joins, you have to add some code to either reload the select music screen or adjust the code to always make both P1/P2 actors and use PlayerJoined to make either player start drawing
